File: src/controller/PriorityQueue.ts
import type { Order } from './Order.js';

export class PriorityQueue {
  private orders: Order[] = [];

  enqueue(order: Order): void {
    // 找到插入位置
    let insertIndex = this.orders.length;
    for (let i = 0; i < this.orders.length; i++) {
      const current = this.orders[i];
      // VIP 订单总是在普通订单前面
      if (order.type === 'vip' && current.type === 'normal') {
        insertIndex = i;
        break;
      }
      // 同类型订单按创建时间排序
      if (order.type === current.type && order.createdAt < current.createdAt) {
        insertIndex = i;
        break;
      }
    }
    this.orders.splice(insertIndex, 0, order);
  }

  dequeue(): Order | undefined {
    return this.orders.shift();
  }

  reinsert(order: Order): void {
    this.enqueue(order);
  }

  get size(): number {
    return this.orders.length;
  }

  getAll(): Order[] {
    return [...this.orders];
  }

  getVipOrders(): Order[] {
    return this.orders.filter(o => o.type === 'vip');
  }

  getNormalOrders(): Order[] {
    return this.orders.filter(o => o.type === 'normal');
  }
}

File: se-take-home-assignment/internal/queue/priority_queue.go
package queue

import (
	"McDonald/internal/model"
)

// PriorityQueue 优先级队列
// VIP 订单排在所有普通订单之前
type PriorityQueue struct {
	orders []*model.Order
}

// NewPriorityQueue creates a new priority queue
func NewPriorityQueue() *PriorityQueue {
	return &PriorityQueue{
		orders: make([]*model.Order, 0),
	}
}

// Enqueue adds an order to the queue
// VIP orders are placed before all Normal orders
// If multiple VIP orders exist, new VIP orders go behind existing VIP orders
func (q *PriorityQueue) Enqueue(order *model.Order) {
	if order.Type == model.OrderTypeVIP {
		// Find the position of the first Normal order
		insertIdx := len(q.orders)
		for i, o := range q.orders {
			if o.Type == model.OrderTypeNormal {
				insertIdx = i
				break
			}
		}
		// Insert at insertIdx
		q.orders = append(q.orders[:insertIdx], append([]*model.Order{order}, q.orders[insertIdx:]...)...)
	} else {
		// Normal orders go to the end
		q.orders = append(q.orders, order)
	}
}

// Dequeue removes and returns the first order from the queue
func (q *PriorityQueue) Dequeue() *model.Order {
	if len(q.orders) == 0 {
		return nil
	}
	order := q.orders[0]
	q.orders = q.orders[1:]
	return order
}

// Peek returns the first order without removing it
func (q *PriorityQueue) Peek() *model.Order {
	if len(q.orders) == 0 {
		return nil
	}
	return q.orders[0]
}

// Remove removes a specific order from the queue
func (q *PriorityQueue) Remove(order *model.Order) bool {
	for i, o := range q.orders {
		if o.ID == order.ID {
			q.orders = append(q.orders[:i], q.orders[i+1:]...)
			return true
		}
	}
	return false
}

// Len returns the number of orders in the queue
func (q *PriorityQueue) Len() int {
	return len(q.orders)
}

// InsertAtEnd inserts a Normal order at the end (for returnToPending scenario)
func (q *PriorityQueue) InsertAtEnd(order *model.Order) {
	q.orders = append(q.orders, order)
}

// GetOrders returns all orders (for debugging)
func (q *PriorityQueue) GetOrders() []*model.Order {
	return q.orders
}
